### ISSUE TITLE
Use PEP 525 async generator to improve data and orders client design

### ISSUE THREAD

**Issue by sgillies** [planetlabs/planet-client-python#706]
[PEP525](https://peps.python.org/pep-0525/) added async generators to Python 3.6. An async generator is an `async def` function or method that uses `yield` instead of `return`. Using it for SDK methods that have multiple results gives us a cleaner design that is in harmony with the designs of async Python software like aiohttp and httpx.

Here is a short bit of code that uses the data client and orders client together. We search the catalog for items using DataClient and then use a list of item ids in creating an order with OrdersClient.

```python
async def main():
    """Use DataClient and OrdersClient to search for and order scenes.
    """
    async with Session() as sess:
        data_client = DataClient(sess)

        # A reader might be led to believe that since we're awaiting the results
        # of search(), the return value has all the results materialized in it.
        search_result_iterator = await data_client.search(["PSScene"], filter_config)

        # But it does not. We still need to asynchronously iterate.
        item_ids = [item["id"] async for item in search_result_iterator]

        client = OrdersClient(sess)
        order_request = {
            "name": "my_order",
            "products": [
                {
                    "item_ids": item_ids,
                    "item_type": "PSScene",
                    "product_bundle": "analytic_udm2",
                }
            ]
        }
        order = await client.create_order(order_request)

        # Wait for order to be completed.
        order_id = order["id"]
        state = await client.wait(order_id, state="success")
        ...
```

To reiterate what I wrote in a comment, a reader of the code might be led to believe that since we're awaiting the results of search(), the return value has all the results already materialized in it. And requiring further iteration then looks confusing. I think it looks odd to await a method call and then asynchronously iterate over the return value. And Python doesn't require it. Neither httpx's [aiter_bytes](https://www.python-httpx.org/async/#streaming-responses) nor aiohttp's [iter_chunked](https://docs.aiohttp.org/en/stable/client_quickstart.html#streaming-response-content) return chunkers. They are async generators.

As an improvement, I propose to make DataClient.search an async generator and allow more concise usage like this:

```python
async def main():
    """Use DataClient and OrderClient to search for and order scenes.
    """
    async with Session() as sess:
        data_client = DataClient(sess)
        item_ids = [item["id"] async for item in data_client.search(["PSScene"], filter_config)]
        ...
```

The new implementation would be almost trivial. It would mean, for DataClient, changing the code at 
 https://github.com/planetlabs/planet-client-python/blob/main/planet/clients/data.py#L185 from

```python
        return Items(request, self._do_request, limit=limit)
```

to

```python
        async for item in Items(request, self._do_request, limit=limit):
            yield item
```

This is not unlike what httpx does here: https://github.com/encode/httpx/blob/8152c4facd0f71e0f376287e41a0810a60fec9c6/httpx/_models.py#L921. It *does* have a chunker, but it's used internally, not by httpx users.

What makes the implementation not trivial is updating tests. That's where most of the work is. I think it's worth doing. 

To support less common situations where one really needs a reference to the search results generator, say to interleave calls to `next(search_results)` with other function calls, one can assign `search_results = client.search()` (without `await`) and then use the awaitable next method of the async generator.

```python
import asyncio

async def func():
    for i in range(10):
        yield i

async def main():
    gen = func()
    print("first")
    # In Python 3.10 this is: print(await anext(gen))
    print(await gen.__anext__())
    print("second")
    print(await gen.__anext__())

asyncio.run(main())
first
0
second
1
```

**Comment by jreiberkyle** [comment]
ahh LOL this is already a ticket. Well, then maybe just summarize the work at the top with the reasoning below it.

**Comment by sgillies** [comment]
@jreiberkyle I broke out the different tasks for this. At the end of the initial comment.

**Comment by jreiberkyle** [comment]
I agree fully with changing things so we no longer return the Orders/Subscriptions/Items/... classes. In fact, as I am in the code working on #739 (a part of which is making changes to `models.Paged`), I am considering changing the code so that `Session` handles paging with a method with a signature something like
```python
def page(response, links_key='_links', next_key = 'next', items_key='items') -> typing.AsyncGenerator[dict, None]
```
and the `Paged` object is instantiated and handled within `page` and the `Paged` class is moved to `http` as a private class that is exclusively used by `Session`.

thoughts?

**Comment by jreiberkyle** [comment]
This change is driven by it feeling kind of awkward passing `Session.request()` as a callback to `Paged` within the clients.

**Comment by sgillies** [comment]
@jreiberkyle how about as a post-2.0 step?

**Comment by jreiberkyle** [comment]
@sgillies LOL you are absolutely right. Thank you for the invitation to continue to remain focused on release-critical work. Ticketed at #757

**Comment by jreiberkyle** [comment]
As I am implementing this I am noticing that this change has the unfortunate byproduct of not actually running the command until the generator is exercised. This has the effect of: the user would think the command was successful since no exception was raised (e.g. `items = run_search()`) but then the error is thrown when they actually exercise the generator. This is an undesirable impact to usability. I will think about a solution that keeps the immediate feedback while shielding the user from implementations such as `Item`, etc.

**Comment by sgillies** [comment]
@jreiberkyle it's in the nature of Python generators to be lazy. I'm interested to see what you come up with!

**Comment by jreiberkyle** [comment]
The thing I am having a hard time with is the naming of the functions.
Some of the functions that deal with paged results are `list_orders`, `run_search`, and `search`. Based on the names, I find it counterintuitive that the results aren't already populated - this is true whether the first page is retrieved (so some of the results are populated) or not (none of them are).

Can you think of a naming convention that is closer to what aiohttp and httpx do, aka `aiter_bytes` and `aiter_chunks`? With those names, I'm not surprised when I receive an iterator that isn't already populated. `aiter_run_search_results()? That's a long hame 😆

**Comment by jreiberkyle** [comment]
Alternatively, we could do less for the user without their express involvement.

[boto3](https://boto3.amazonaws.com/v1/documentation/api/latest/guide/paginators.html#creating-paginators) requires the user to specifically create a paginator and page through

excerpt:
```python
import boto3

# Create a client
client = boto3.client('s3', region_name='us-west-2')

# Create a reusable Paginator
paginator = client.get_paginator('list_objects')

# Create a PageIterator from the Paginator
page_iterator = paginator.paginate(Bucket='my-bucket')

for page in page_iterator:
    print(page['Contents'])
```

**Comment by jreiberkyle** [comment]
Maybe something like `aiter_search` instead of `search` and `aiter_list_orders` instead of `list_orders`? This is my current leaning - to implement the changes you recommend here and also prepend `aiter_` to the function names.

**Comment by jreiberkyle** [comment]
I've decided on adding `_aiter` to the end of the function names. So `run_search` becomes `run_search_aiter` and the docstring goes from "Execute a saved search" to "Iterate over results from a saved search"
